wallet: bugfix, invalid CoinsResult cached total amount

furszy opened the pull request:
This comes with #26559.

Solving few bugs inside the wallet's transaction creation
process and adding test coverage for them.
Plus, making use of the `CoinsResult::total_amount` cached value
inside the Coin Selection process to return early if we don't have
enough funds to cover the target amount.

### Bugs

1) The `CoinsResult::Erase` method removes only one
output from the available coins vector (there is a [loop break](https://github.com/bitcoin/bitcoin/blob/c1061be14a515b0ed4f4d646fcd0378c62e6ded3/src/wallet/spend.cpp#L112)
that should have never been there) and not all the preset inputs.

   Which on master is not a problem, because since [#25685](https://github.com/bitcoin/bitcoin/pull/25685)
   we are no longer using the method. But, it's a bug on v24
   (check [#26559](https://github.com/bitcoin/bitcoin/pull/26559)).

   This method it's being fixed and not removed because I'm later using it to solve
   another bug inside this PR.

2) As we update the total cached amount of the `CoinsResult` object inside
   `AvailableCoins` and we don't use such function inside the coin selection
   tests (we manually load up the `CoinsResult` object), there is a discrepancy
   between the outputs that we add/erase and the total amount cached value.

### Improvements

* This makes use of the `CoinsResult` total amount field to early return
  with an "Insufficient funds" error inside Coin Selection if the tx target
  amount is greater than the sum of all the wallet available coins plus the
  preset inputs amounts (we don't need to perform the entire coin selection
  process if we already know that there aren't enough funds inside our wallet).

### Test Coverage

1) Adds test coverage for the duplicated preset input selection bug that we have in v24.
  Where the wallet invalidly selects the preset inputs twice during the Coin Selection
  process. Which ends up with a "good" Coin Selection result that does not cover the
  total tx target amount. Which, alone, crashes the wallet due an insane fee.
  But.. to make it worst, adding the subtract fee from output functionality
  to this mix ends up with the wallet by-passing the "insane" fee assertion,
  decreasing the output amount to fulfill the insane fee, and.. sadly,
  broadcasting the tx to the network.

2) Adds test coverage for the `CoinsResult::Erase` method.

furszy committed:
wallet: bugfix, 'CoinsResult::Erase' is erasing only one output of the set

The loop break shouldn't have being there.

furszy committed:
test: bugfix, coinselector_test, use 'CoinsResult::Erase/Add' instead of direct member access

Aside from the cleanup, this solves a bug in the following-up commit. Because, in these
tests, we are manually adding/erasing outputs from the CoinsResult object but never
updating the internal total amount field.

furszy committed:
wallet: SelectCoins, return early if wallet's UTXOs cannot cover the target

The CoinsResult class will now count the raw total amount and the effective
total amount internally (inside the 'CoinsResult::Add' and 'CoinsResult::Erase'
methods).
So there is no discrepancy between what we add/erase and the total values.
(which is what was happening on the coinselector_test because the 'CoinsResult'
object is manually created there, and we were not keeping the total amount
in sync with the outputs being added/removed).

furszy committed:
test: wallet, coverage for CoinsResult::Erase function

furszy committed:
test: Coin Selection, duplicated preset inputs selection

This exercises the bug inside CoinsResult::Erase that
ends up on (1) a wallet crash or (2) a created and
broadcasted tx that contains a reduced recipient's amount.

This is covered by making the wallet selects the preset
inputs twice during the coin selection process.

Making the wallet think that the selection process result covers
the entire tx target when it does not. It's actually creating
a tx that sends more coins than what inputs are covering for.

Which, combined with the SFFO option, makes the wallet
incorrectly reduce the recipient's amount by the difference
between the original target and the wrongly counted inputs.
Which means, a created and relayed tx sending less coins to
the destination than what the user inputted.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -102,15 +102,19 @@ void CoinsResult::Clear() {
+            if (it == coins_to_remove.end()) return false;
+
+            // remove coin
+            coins_to_remove.erase(it);
```
In 7f6abb39159cd55b6c9193630ca4c3d9ae551be6 "wallet: bugfix, 'CoinsResult::Erase' is erasing only one output of the set"

How come we need to have `coins_to_remove` and erase the removed coins from it? `coins_to_remove` isn't being used anywhere else so this doesn't seem to be doing anything?

furszy replied:
It's to decrease the search time while moving across the `CoinsResult` vectors.

`Erase` receives a const set with all the coins that needs to be removed from the `CoinsResult` internal vectors. So, what we do is loop across each coins vector and, for each coin inside them, **we test whether the coin is inside the set of coins provided by the caller**, if it's there, then we update the internal state and return true so `std:.remove_if` does its magic.
(bolded the important part)

So, if we remove coins as we find them on the `CoinsResult` vectors from the to_remove set, we will continually search inside a container with less elements up until there is none and we can stop the search earlier.

achow101 replied:
I don't think that's really a useful optimization since `outs` is likely to be very small. If we are concerned about the complexity of `find`, then I think it would be better to use an `unordered_set` rather than erasing after each removal.

This isn't really a blocking comment, I just found it a bit confusing.

furszy replied:
yeah, an `unordered_set` sounds good too.

still, what I wanted to add (which I actually didn't add it for some reason..) is the capability of stopping early if the set has no elements. E.g. if the wallet has 100k UTXO inside, the preset inputs set is small, and the coins appear early in the loop, the process could stop right away (no need to keep traversing the entire available coins vector if we already found them).

achow101 reviewed (commented):
Good catch! Unfortunate that this was found right as we're doing the 24.0 release.

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -575,6 +584,14 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
+    // Return early if we cannot cover the target with the wallet's UTXO.
+    // We use the total effective value if we are not subtracting fee from outputs and 'available_coins' contains the data.
+    CAmount available_coins_total_amount = coin_selection_params.m_subtract_fee_outputs ? available_coins.total_amount :
+            (available_coins.total_effective_amount.has_value() ? *available_coins.total_effective_amount : available_coins.total_amount);
```
I don't think it's correct to fallback on `available_coins.total_amount` when we don't do SFFO.

with SFFO we must have effective amount, if we don't - something went wrong.

furszy replied:
Hmm, need a bit of context here.
Guess that you are saying that because, with SFFO, the wallet needs to select a group of UTXOs that will be able to pay for themselves? So, the entire coin selection relies on the effective value existence so.. if it's not there there is a bug?

achow101 replied:
By the time we get here, if `total_effective_amount` doesn't have a value, that means that none of the UTXOs in this wallet were solvable, so we can't do any coin selection here in that case. So I think it would be better to always return early in that case.

furszy replied:
> By the time we get here, if total_effective_amount doesn't have a value, that means that none of the UTXOs in this wallet were solvable, so we can't do any coin selection here in that case

Hmm, code wise, that doesn't seems to be what is happening.

At the `COutput` constructor, if the coin isn't solvable then `effective_value` will be equal to the nominal output value. So it will always have an effective value during the tx creation process.

The only way that a `COutput` has no `effective_value` is when we don't provide the feerate param to `AvailableCoins` (which happens on all the flows that uses `AvailableCoins` that are not part of the tx creation process).

achow101 replied:
Oh, hmm, I thought it checked `input_bytes` as well.

In that case, `total_effective_amount` should always have a value if we have any coins. If it isn't set, then we don't have any coins and we can still return early.

S3RK commented on src/wallet/spend.h:
```diff
@@ -51,8 +51,10 @@ struct CoinsResult {
+    /** Sum of all available coins raw value */
     CAmount total_amount{0};
+    /** Sum of all available coins effective value (each output value minus fees required to spend it) */
+    std::optional<CAmount> total_effective_amount{0};
```
note to self: not a huge fan of how this looks like with `optional`. The code would look much simpler if we can just always pass a fee rate to `AvailableCoins` (we can set it 0 by default). Can't make my mind which way is better...

furszy replied:
I would agree if we were using `AvailableCoins` only for the tx creation process, but we are using the function for other processes that don't require to calculate the effective value nor the output spending fee.

We definitely can improve it but I don't think that this PR is the best place to do it.

Side from that, from my side, have some plans to re-write the `AvailableCoins` functionality differently post #25806 (because, after it, will be able to continually connect outputs to existent output groups, outside of the coin selection process which is currently forcing us to do the whole calculation on-demand).

S3RK reviewed (commented):
oh wow, that's a nasty bug. 

Almost Code Review ACK modulo use of `available_coins.total_amount` with SFFO in the last commit.

I was quite surprised something like this wasn't caught by some assert. So I dig up and maybe we can add something like e783af37a77bbe91cd6ec8802292838377d9a0aa while you're at it

Sjors commented:
Asking here instead of in the backport: is this only in v24 or also earlier versions?

fanquake commented:
> Asking here instead of in the backport: is this only in v24 or also earlier versions?

My understanding is only v24.x. Seems like the buggy behaviour was introduced in #24584 + #25734?

furszy commented:
> Asking here instead of in the backport: is this only in v24 or also earlier versions?

Yeah. @Sjors only for v24. The dangerous bug was introduced in #25734 which is only on v24, and after branching off, we merged #25685 which fixed it.

furszy commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,63 @@ def test_add_inputs_default_value(self):
+        self.nodes[0].sendmany("", outputs)
+        self.generate(self.nodes[0], 1)
+
+        # Select the preset inputs and lock the rest of the coins
```
let me double check it, IIRC I added it to make the wallet have enough funds in case we have an early balance check in the flow.

furszy replied:
All good, just removed it. Seems that I was just over-thinking it.

furszy commented on src/wallet/test/spend_tests.cpp:
```diff
@@ -112,5 +112,38 @@ BOOST_FIXTURE_TEST_CASE(FillInputToWeightTest, BasicTestingSetup)
     // Note: We don't test the next boundary because of memory allocation constraints.
 }
 
+BOOST_FIXTURE_TEST_CASE(wallet_duplicated_preset_inputs_test, TestChain100Setup)
```
yeah. I almost remove it but.. ended up leaving because (1) the severity of the bug and (2) because it's easier to debug and assert that the code is failing where is supposed to be failing (in the functional test case, as we use the `send` RPC command, we do have other stuff involved).

I'm fine either way, if this still doesn't convince you and more people don't see it useful, can remove it for sure.

josibake replied:
I also agree that the second unit test (which isn't really a unit test) seems redundant and also a little confusing to me. I think the unit test for `CoinsResult::Erase` in an earlier commit is correct for a unit test and the functional test is better suited for testing the wallet behavior

furszy commented on src/wallet/test/spend_tests.cpp:
```diff
@@ -112,5 +112,38 @@ BOOST_FIXTURE_TEST_CASE(FillInputToWeightTest, BasicTestingSetup)
+    auto available_coins = AvailableCoins(*wallet);
+    std::vector<COutput> coins = available_coins.All();
+    std::set<COutPoint> preset_inputs = {coins[0].outpoint, coins[1].outpoint};
+    // Lock the rest of the coins
```
not really, removed it. Thanks

furszy commented:
Feedback tackled, thanks @S3RK and @achow101.

Changes:
1) In the tests, removed the unneeded coins lock.
2) Now `SelectCoins` returns early if the wallet's available coins `total_effective_amount` is not available (it does not fallback to use the raw total amount anymore).

luke-jr commented:
I think probably `CoinsResult::coins` should be a `std::map<OutputType, std::unordered_set<COutput>>` instead of [map of] vectors. It's too strange that we could end up with the same output in there twice. Maybe even an assert failure if we try to add the same ones twice?

luke-jr reviewed (changes requested):
The bugfix should also be done in its own commit, absent any refactoring.

I think this means just dropping the `break` and replacing the `find_if` with `remove_if`?

josibake commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,59 @@ def test_add_inputs_default_value(self):
+            "add_to_wallet": False
+        }
+
+        # First case, use 'subtract_fee_from_outputs' to make the wallet create a tx that pays an insane fee on v24.0.
+        assert_raises_rpc_error(-4, "Insufficient funds", wallet.send, outputs=[{wallet.getnewaddress(address_type="bech32"): 14}], options=options)
+
+        # Second case, don't use 'subtract_fee_from_outputs' to make the wallet crash due the insane fee on v24.0
+        del options["subtract_fee_from_outputs"]
+        assert_raises_rpc_error(-4, "Insufficient funds", wallet.send, outputs=[{wallet.getnewaddress(address_type="bech32"): 14}], options=options)
```
As it's written, the test comment doesn't make sense unless you have the background context. The comment describes a certain type of failure (insane fee, or wallet crash), but the assertion seems to check something entirely different. Perhaps it's worth writing a long comment which explains why you are checking for insufficient funds and what you would expect if the insufficient funds assert fails.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -106,7 +106,13 @@ void CoinsResult::Erase(const std::unordered_set<COutPoint, SaltedOutpointHasher
         auto remove_it = std::remove_if(vec.begin(), vec.end(), [&](const COutput& coin) {
-            return coins_to_remove.count(coin.outpoint) == 1;
+            // remove it if it's on the set
+            if (coins_to_remove.count(coin.outpoint) == 0) return false;
```
in https://github.com/bitcoin/bitcoin/pull/26560/commits/1b14bd7ac77027351a1e5b3ef21fb2d1de659efd:

same comment from before, I'd suggest using `find` here.

josibake replied:
also, might be worth adding the `if () return false;` in the earlier commit so you don't end up touching this line twice?

furszy replied:
> in [1b14bd7](https://github.com/bitcoin/bitcoin/commit/1b14bd7ac77027351a1e5b3ef21fb2d1de659efd):
> 
> same comment from before, I'd suggest using `find` here.

answered in the first commit comment https://github.com/bitcoin/bitcoin/pull/26560#discussion_r1034897955


> also, might be worth adding the `if () return false;` in the earlier commit so you don't end up touching this line twice?

not sure what you meant, the first commit has a single line (the `count` function will either return 1 if the element is in the set or 0 if not).
```c++
return coins_to_remove.count(coin.outpoint) == 1;
```

josibake replied:
yeah, more just saying you could write the first commit like:

```
if (coins_to_remove.count(coin.outpoint) == 0) return false;
return true;
```

because you know that in a later commit you will add the `total_amount` and `total_effective` code, and this way you avoid re-writing the same line twice. it doesn't really matter in this case, just a suggestion. in general, it can be confusing as a reviewer when the same line of code is edited multiple times in different commits

furszy replied:
Ideally, commits should introduce atomic changes. Did the first one in this way so it can be backported without any dependency into a different branch (due the lack of https://github.com/bitcoin/bitcoin/pull/25685, v24 branch cannot cherry-pick https://github.com/bitcoin/bitcoin/commit/1b14bd7ac77027351a1e5b3ef21fb2d1de659efd).
So, we would only have that commit inside v24, which I think that is ugly.

josibake reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26560/commits/1b14bd7ac77027351a1e5b3ef21fb2d1de659efd

Glad you caught this! Surprised and pretty bummed this slipped through in #25734 :disappointed: 

Left a few suggestions, but non-blocking.

furszy commented:
> I think probably `CoinsResult::coins` should be a `std::map<OutputType, std::unordered_set<COutput>>` instead of [map of] vectors. It's too strange that we could end up with the same output in there twice. Maybe even an assert failure if we try to add the same ones twice?

@luke-jr: we can't end up with the same output twice there. That struct is loaded inside `AvailableCoins` which traverses the wallet's txes map once.
In v24, where we don't have https://github.com/bitcoin/bitcoin/pull/25685 merge, the preset inputs were fetched twice: first inside `AvailableCoins` which was appending them to the `CoinsResult` vectors and secondly inside `SelectCoins` where we were manually fetching them from the wallet txes map [here](https://github.com/bitcoin/bitcoin/blob/c1061be14a515b0ed4f4d646fcd0378c62e6ded3/src/wallet/spend.cpp#L542-L587) (by calling `GetWalletTx(outpoint)`). 
Then, `SelectCoins` was removing the preset inputs from the `CoinsResult` vectors before start the automatic coin selection process so the process does not select them (bug was here, the `Erase` function was not removing them). Then, at the end of the selection process, we were merging both results (the preset inputs selection result and the automatic coin selection result).

furszy commented on src/wallet/spend.cpp:
```diff
@@ -102,15 +102,13 @@ void CoinsResult::Clear() {
-        }
+    for (auto& [type, vec] : coins) {
+        auto remove_it = std::remove_if(vec.begin(), vec.end(), [&](const COutput& coin) {
+            return coins_to_remove.count(coin.outpoint) == 1;
```
`coins_to_remove` is an unordered set, not a vector. Internally, uses a hash table where keys are hashed into indices. So, time complexity wise, `count` and `find` should take the same constant time.

josibake replied:
ah, good point! i forgot how `find` \ `count` work is dependent on the underlying container

furszy commented on src/wallet/spend.h:
```diff
@@ -47,7 +47,7 @@ struct CoinsResult {
     size_t Size() const;
     void Clear();
-    void Erase(std::set<COutPoint>& preset_coins);
+    void Erase(const std::unordered_set<COutPoint, SaltedOutpointHasher>& outs);
```
yeah, thanks

josibake commented:
fwiw, I ran the failing CI test locally and it passed, so I think all you need to do is restart the test

S3RK committed:
wallet: add assert to SelectionResult::Merge for safety

furszy commented:
Updated per feedback, thanks everyone. [Small diff](https://github.com/bitcoin/bitcoin/compare/1b14bd7ac77027351a1e5b3ef21fb2d1de659efd..8f9bd177255f285a53871fa278528338407af267).

Changes:
1) Per @S3RK feedback, cherry-picked a simple fallback assertion.
   I still think that we might want to merge different `SelectionResult` objects that share some outputs in the future but.. now it's definitely not the right time to do it.
2) Per @josibake feedback, added [#26559](https://github.com/bitcoin/bitcoin/pull/26559) mention in the functional test cases and few more details in the comments.

achow101 commented:
ACK eb76557331b4f34b568da1d9c589c7005d7ab46d

josibake commented:
ACK https://github.com/bitcoin/bitcoin/commit/eb76557331b4f34b568da1d9c589c7005d7ab46d

reviewed the diff, looks good

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -106,7 +106,13 @@ void CoinsResult::Erase(const std::unordered_set<COutPoint, SaltedOutpointHasher
+
+            // update cached amounts
+            total_amount -= coin.txout.nValue;
+            if (coin.HasEffectiveValue()) total_effective_amount = *total_effective_amount - coin.GetEffectiveValue();
```
Are we sure that `total_effective_value` is always set here?

Maybe `assume(total_effective_amount.has_value()` here?

S3RK replied:
I had this question myself. I figured if we remove the coin with effective value, then the coin was added before, hence we have `total_effective_value` defined. Please correct me if my reasoning is flawed.

furszy replied:
yeah. As we only use `CoinsResult::Add` to append new coins: if we have the coin inside `CoinsResult`, and the coin has an effective value, then we must have `total_effective_amount`.

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -575,6 +584,14 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
 
+    // Return early if we cannot cover the target with the wallet's UTXO.
+    // We use the total effective value if we are not subtracting fee from outputs and 'available_coins' contains the data.
+    CAmount available_coins_total_amount = coin_selection_params.m_subtract_fee_outputs ? available_coins.total_amount :
+            (available_coins.total_effective_amount.has_value() ? *available_coins.total_effective_amount : 0);
```
We will never use `SelectCoins()` without a feerate. If you end up not having total_effective_amount here, something went terribly wrong. Perhaps just:

```suggestion
    assume(available_coins.total_effective_amount.has_value());
    CAmount available_coins_total_amount = coin_selection_params.m_subtract_fee_outputs ? available_coins.total_amount : *available_coins.total_effective_amount);
```

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -452,12 +452,16 @@ void SelectionResult::AddInputs(const std::set<COutput>& inputs, bool subtract_f
 
 void SelectionResult::Merge(const SelectionResult& other)
 {
+    // Obtain the expected selected inputs count after the merge (for now, duplicates are not allowed)
```
```suggestion
    // Store sum of combined input sets to check that no UTXO was used twice
```

furszy replied:
this relies on https://github.com/bitcoin/bitcoin/pull/26560#discussion_r1038307694 (where we are talking about whether should include the assertion commit here or in a follow-up PR).
Please, feel free to comment there.

murchandamus reviewed (commented):
ACK eb76557331b4f34b568da1d9c589c7005d7ab46d

achow101 commented on src/wallet/spend.h:
```diff
@@ -47,12 +47,14 @@ struct CoinsResult {
+    /** Sum of all available coins raw value */
     CAmount total_amount{0};
+    /** Sum of all available coins effective value (each output value minus fees required to spend it) */
+    std::optional<CAmount> total_effective_amount{0};
```
I think it would be useful to make `total_amount`, `total_effective_amount`, and `coins` private members to ensure that we are only adding/removing things with `Add` and `Erase`.

furszy replied:
done

S3RK commented:
ACK eb76557331b4f34b568da1d9c589c7005d7ab46d

Will reack if you address nits

glozow commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,67 @@ def test_add_inputs_default_value(self):
+            ],
+            "add_inputs": True,
+            "subtract_fee_from_outputs": [0],
+            "add_to_wallet": False
```
Some questions while trying to reproduce the "insane fee on v24.0" case:

- Why is `add_to_wallet: False` necessary? Based on the comment, I assumed the test would demonstrate the wallet actually spending fees.
- What impact does locking the inputs have? Deleting it didn't seem to change the fee, though maybe I'm missing something.

glozow commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,67 @@ def test_add_inputs_default_value(self):
+    def test_preset_inputs_selection_crash(self):
+        self.log.info('Test wallet preset inputs selection')
+
+        # This exercises the bug presented in #26559 that
+        # ended up on (1) a wallet crash or (2) a created and
+        # broadcasted tx that pays an insane fee amount.
+        #
+        # This is covered by making the wallet selects the preset
+        # inputs twice during the coin selection process (which
+        # is a dangerous bug on v24.0).
+        #
+        # Making the wallet think that the selection process result covers
+        # the entire tx target when it does not. It's actually creating
+        # a tx that sends more coins than what inputs are covering for.
+        #
+        # Which, combined with the subtract fee from outputs option,
+        # makes the wallet generate and broadcast a tx that pays
+        # up to the sum of all the preset inputs, minus one, amounts in
+        # fee :(.
```
If I were reading this a year from now (e.g. because I reintroduced the bug and this failed), I would need more help understanding what the test is doing. This comment is a bit too focused on the user effects of the bug (which is helpful but the commit message + PR description already says the same thing); the test should instead describe what unexpected code behavior it's aiming to catch.

```suggestion
        # Check that the wallet handles preset inputs properly,
        # i.e. rmoves them from the coin selection UTXO pool
        # and accurately accounts for the fee and value of all
        # inputs. Amounts should not be double-counted and
        # the wallet should not crash.
```

glozow commented on src/wallet/coinselection.cpp:
```diff
@@ -452,12 +452,16 @@ void SelectionResult::AddInputs(const std::set<COutput>& inputs, bool subtract_f
         m_algo = other.m_algo;
     }
     util::insert(m_selected_inputs, other.m_selected_inputs);
+    assert(m_selected_inputs.size() == expected_count);
```
It's a bit unclear to me how the assert added in eb76557331b4f34b568da1d9c589c7005d7ab46d is more safe? Wouldn't this mean we still crash if we haven't caught a similar bug before shipping?

glozow replied:
I suggest removing this last commit for this PR, and following up with an "add sanity checks to coin selection code" PR.

Of course we don't want to create a bad transaction, but an assert would cause the user's node to crash in case we have implemented it wrong. I think something better could be to add `if(!Assume(...)) {return "sorry, something went wrong and transaction creation failed. please report this bug...";}` which should stop the wallet from doing something horrible but also not crash. Apologies if this is bikesheddy - my concrete suggestion is to remove this for now, and do a followup PR for these types of changes. Other appropriate sanity checks could include `AddInput()` never adds an input that's already there, `SelectCoins` result should cover `selection_target`, etc.

furszy replied:
> Of course we don't want to create a bad transaction, but an assert would cause the user's node to crash in case we have implemented it wrong. I think something better could be to add if(!Assume(...)) {return "sorry, something went wrong and transaction creation failed. please report this bug...";} which should stop the wallet from doing something horrible but also not crash.

Sure. I totally agree on using more `Assume` + user friendly error return messages than crashing the node with a bare assertion. 

I do actually implemented the capability to return error messages from the Coin Selection process inside [#25806](https://github.com/bitcoin/bitcoin/pull/25806) -> 45bc10687f7b0131fd1c4575406b62f0608e2c01 few months ago (which is later used to return specific error messages instead of the current general "Insufficient Funds". Which is required to implement your suggestions as we currently don't have a way to return specific errors to the upper layers).

So, in short, could drop the last commit and open a follow-up PR moving forward with it, and more, for sure, but.. as I added the last commit by reviewers request, would be good to see if @S3RK is sync with this reasoning too (otherwise we would be going in circles).
Note: Since, after the fix, this problem cannot realistically happen anymore, I'm fine to move either way.

achow101 replied:
I think it's find to change this to `Assume` for this PR. We should probably try to avoid asserts in general.

S3RK replied:
I'll be fine with any of the following: keep asserts as I proposed, replaced them with `Assume` for this PR, remove the commit and address this in a follow up. With that said, here's my strong opinion held lightly.

In the absence of ability of returning error messages from coin selection we have only two options. So the question is what is better within the scope of this PR:
a) crash the node if assert is hit
b) keep executing and potentially creating bad transactions.

Both are bad, but a) is strictly better, because recovering from a crash is much easier than from a bad tx. Replacing `assert`s and returning error messages could/should done in a follow up.

Now I agree that in general the wallet should not ever crash the node, but for this we need a general mechanism, something like multiprocess.

Also, my original suggestion included check within `AddInput()` and `AddInputs()`, as `SelectionResult` is not designed to work with duplicated inputs. This would make the class safe and its interface consistent.

MarcoFalke replied:
Another alternative (not sure if applicable here) would be to throw an exception. This should only crash when running in the gui, but not when called from the RPC server.

josibake replied:
It seems we all (mostly) agree that the `assert` is not ideal and how to deal with this should be addressed in a follow-up with more discussion, so I'd recommend dropping the commit so we don't merge code that we know will be removed/refactored shortly after and to avoid any confusion when backporting commits from this PR for v24

furszy replied:
working on it 🚜 .

fanquake commented on src/wallet/test/spend_tests.cpp:
```diff
@@ -112,5 +112,34 @@ BOOST_FIXTURE_TEST_CASE(FillInputToWeightTest, BasicTestingSetup)
+        coin_control.Select(outpoint);
+    }
+
+    // First case, use 'subtract_fee_from_outputs' to make the wallet create a tx that pays an insane fee on v24.0.
```
I don't think we should be adding specific version numbers to comments, because this just becomes (mostly) incorrect, once we've merged a backport, and doesn't really provide any value otherwise.
```suggestion
    // First case, use 'subtract_fee_from_outputs' to make the wallet create a tx that pays an insane fee.
```

furszy replied:
done

fanquake commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,67 @@ def test_add_inputs_default_value(self):
+        }
+
+        # First case, use 'subtract_fee_from_outputs' to make the wallet create a tx that pays an insane fee on v24.0.
+        # (further details in #26559).
```
There is no need to continually write `v24.0` and `(further details in #26559)` (here and below and further above). If you need to provide further context or details, please put them in the commit (or PR) descriptions, so they are available with the code change.

furszy replied:
those were my initial thought actually, added this on the last push per request https://github.com/bitcoin/bitcoin/pull/26560#discussion_r1034800106.

seems that not everyone read the PR/commit description. But well, will just revert it.

josibake replied:
Sorry if this wasn't clear in my comment, but I wasn't suggesting adding `v24` and links to the PR. My comment was to point out that the comments in the test didn't match what was being checked in the assertion, which would be confusing to someone who didn't have the background context. I understand this is a bit tricky because the test is for a bug that is no longer there, but ideally, the comment should explain why you are asserting "Insufficient Funds" in the happy case, and why, if the bug were present, that you would expect the "Insufficient Funds" assertion to fail. If you do want to link to the PR for more context, perhaps a top-level comment would be the best place?

furszy replied:
Ok, done. Ready for another review round.

glozow commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,67 @@ def test_add_inputs_default_value(self):
         self.nodes[2].unloadwallet("test_preset_inputs")
 
+    def test_preset_inputs_selection_crash(self):
+        self.log.info('Test wallet preset inputs selection')
+
+        # This exercises the bug presented in #26559 that
+        # ended up on (1) a wallet crash or (2) a created and
+        # broadcasted tx that pays an insane fee amount.
+        #
+        # This is covered by making the wallet selects the preset
+        # inputs twice during the coin selection process (which
+        # is a dangerous bug on v24.0).
+        #
+        # Making the wallet think that the selection process result covers
+        # the entire tx target when it does not. It's actually creating
+        # a tx that sends more coins than what inputs are covering for.
+        #
+        # Which, combined with the subtract fee from outputs option,
+        # makes the wallet generate and broadcast a tx that pays
+        # up to the sum of all the preset inputs, minus one, amounts in
+        # fee :(.
+
+        # Create and fund the wallet with 2 UTXO of 5 BTC each
```
```suggestion
        self.log.info('Test wallet preset inputs are not double-counted or reused in coin selection')

        # Create and fund the wallet with 2 UTXO of 5 BTC each, total balance 10 BTC.
```

furszy replied:
done

glozow commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,67 @@ def test_add_inputs_default_value(self):
+            "add_to_wallet": False
+        }
+
+        # First case, use 'subtract_fee_from_outputs' to make the wallet create a tx that pays an insane fee on v24.0.
+        # (further details in #26559).
+        assert_raises_rpc_error(-4, "Insufficient funds", wallet.send, outputs=[{wallet.getnewaddress(address_type="bech32"): 14}], options=options)
+
+        # Second case, don't use 'subtract_fee_from_outputs' to make the wallet crash due the insane fee on v24.0.
+        # (further details in #26559).
```
```suggestion
        # Attempt to send 14 BTC. The wallet should exclude the preset inputs from the pool of
        # available coins, realize that there is not enough money to fund the 14 BTC payment,
        # and fail with "insufficient funds."
        # Even with SFFO, the wallet can only afford to send 10 BTC.

        # If the wallet does not properly exclude preset inputs from the pool of available coins
        # prior to coin selection, it may create a transaction that does not fund the full payment
        # amount or, through SFFO, incorrectly reduce the payment amount so that payment+fee is not
        # equal to 14 BTC.
        assert_raises_rpc_error(-4, "Insufficient funds", wallet.send, outputs=[{wallet.getnewaddress(address_type="bech32"): 14}], options=options)
```

furszy replied:
Updated, thanks

glozow commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,67 @@ def test_add_inputs_default_value(self):
+
+        # First case, use 'subtract_fee_from_outputs' to make the wallet create a tx that pays an insane fee on v24.0.
+        # (further details in #26559).
+        assert_raises_rpc_error(-4, "Insufficient funds", wallet.send, outputs=[{wallet.getnewaddress(address_type="bech32"): 14}], options=options)
```
Cherry-picking this test on top of 24.x, I can see that the wallet mistakenly uses a preset input twice and deducts the fee+5BTC from the payment, but the fee is actually exactly what is expected (since at the end the transaction only has 10BTC inputs and not 15): https://github.com/glozow/bitcoin/blob/test-n26560/test/functional/rpc_fundrawtransaction.py#L1250-L1266

@achow101 suggested that maybe `nFeeRet` from `FundTransaction()` or  `CreatedTransactionResult::fee` could be incorrect but I wasn't able to get a test to show that. @achow101 then pointed me to this code showing the sffo fee should be correctly set before `CreateTransaction` returns:
https://github.com/bitcoin/bitcoin/blob/f668a3a85933512e7efc369b5b2922d44b4bad82/src/wallet/spend.cpp#L971-L1004

I could be mistaken, but in that case could you maybe write a test for v24.0 that shows what the expected vs actual fee are? Otherwise I'd suggest removing this from the comments/commit message/PR descriptions.

Having 2 test cases for sffo and non-sffo codepaths is still meaningful, but I think the current comment is misleading and can cause us to have trouble understanding this test in the future.

furszy replied:
Done, updated per feedback. Thanks.

Some context:
The "insane fee" label came from the `nFeeRet` calculation. Which, in case of this bug, sets a high value than is deducted from the recipient's amount. Which.. as we spoke via dm, it's true that this is internal to the wallet only.. so the description should be something like "double-counted preset inputs during Coin Selection incorrectly deduces the user's selected recipient's amount" (or something similar)

glozow reviewed (commented):
strong Concept ACK, I understand the bug has a significant impact and agree with the fix, but the documentation should be changed as I think it is currently incorrect and could be misleading to someone in the future trying to debug / reproduce what was wrong with v24.0.

TLDR I don't think the "insane fee" bug is actually possible? I've made some suggestions for what I think the documentation should be inline (same suggestion wrt the commit messages).

furszy committed:
refactor: make CoinsResult total amounts members private

furszy commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -980,5 +980,43 @@ BOOST_AUTO_TEST_CASE(SelectCoins_effective_value_test)
+        for (int i = 0; i < 2; i++) {
+            outs_to_remove.emplace(coins[i].outpoint);
+        }
+        available_coins.Erase(outs_to_remove);
```
sounds good

furszy commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1200,6 +1201,67 @@ def test_add_inputs_default_value(self):
+
+        # Now let's create the tx
+        options = {
+            "inputs": [
+                {
+                    "txid": preset_inputs[0]["txid"],
+                    "vout": preset_inputs[0]["vout"]
+                },
+                {
+                    "txid": preset_inputs[1]["txid"],
+                    "vout": preset_inputs[1]["vout"]
+                },
+            ],
```
nop, you are right. I was doing this stuff manually at the beginning, can simplify it now. Thanks

furszy commented on src/wallet/test/spend_tests.cpp:
```diff
@@ -112,5 +112,34 @@ BOOST_FIXTURE_TEST_CASE(FillInputToWeightTest, BasicTestingSetup)
+        coin_control.Select(outpoint);
+    }
+
+    // First case, use 'subtract_fee_from_outputs' to make the wallet create a tx that pays an insane fee on v24.0.
```
yeah ok, I used that term because the `nFeeRet` calculation prior to the SFFO process, can go insanely high based on the duplicated preset inputs, then we deduct it from the recipient's target.
will improve the documentation.

glozow reviewed (commented):
ACK cf793846978a8783c23b66ba6b4f3f30e83ff3eb (I assume these 3 are the backport commits) Verified that the tests in cf793846978a8783c23b66ba6b4f3f30e83ff3eb pass with the fix, fail on 24.x without the fix. I now think the documentation is accurate and would be helpful in a situation in which the test has broken again - thanks for taking the feedback!

ACK 7362f8e5e2497bc1ef27bfa871fc6dd306dd33c3, I assume there will be a followup PR to add coin selection sanity checks and we can discuss the best way to do that there.

achow101 commented:
ACK 7362f8e5e2497bc1ef27bfa871fc6dd306dd33c3

josibake reviewed (commented):
ACK [7362f8e](https://github.com/bitcoin/bitcoin/pull/26560/commits/7362f8e5e2497bc1ef27bfa871fc6dd306dd33c3)

Give the discussion, I think it's best if we drop https://github.com/bitcoin/bitcoin/pull/26560/commits/3282fad59908da328f8323e1213344fe58ccf69e and address it in a follow-up, but not worth holding the PR up if the author feels differently.

A big thanks to furszy for incorporating all of the feedback regarding the tests and documentation: everything is very clear now as to what's being tested and why

fanquake commented:
Backported the 3 commits in #26616.

484812 reviewed (commented):
Gj
